apps/heft/src/plugins/DeleteGlobsPlugin.ts

```diff
@@ -0,0 +1,135 @@
+      build.hooks.preCompile.tap(PLUGIN_NAME, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.preCompile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.compile.tap(PLUGIN_NAME, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.compile, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.bundle.tap(PLUGIN_NAME, (bundle: IBundleSubstage) => {
+        bundle.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.bundle, logger, heftConfiguration);
+        });
+      });
+
+      build.hooks.postBuild.tap(PLUGIN_NAME, (postBuild: IPostBuildSubstage) => {
+        postBuild.hooks.run.tapPromise(HEFT_STAGE_TAP, async () => {
+          await this._runDeleteForHeftEvent(HeftEvent.postBuild, logger, heftConfiguration);
+        });
+      });
+    });
+  }
+
+  private async _runDeleteForHeftEvent(
+    heftEvent: HeftEvent,
+    logger: ScopedLogger,
+    heftConfiguration: HeftConfiguration,
+    additionalPathsToDelete?: Set<string>
+  ): Promise<void> {
+    let deletedFiles: number = 0;
+    let deletedFolders: number = 0;
+
+    const eventActions: IHeftEventActions = await ConfigFile.getConfigConfigFileEventActionsAsync(
+      heftConfiguration
+    );
+
+    const pathsToDelete: Set<string> = new Set<string>(additionalPathsToDelete);
+    for (const deleteGlobsEventAction of eventActions.deleteGlobs.get(heftEvent) || []) {
+      for (const globPattern of deleteGlobsEventAction.globsToDelete) {
+        const resolvedPaths: string[] = await this._resolvePathAsync(
+          globPattern,
+          heftConfiguration.buildFolder
+        );
+        for (const resolvedPath of resolvedPaths) {
+          pathsToDelete.add(resolvedPath);
+        }
+      }
+    }
+
+    await Async.forEachLimitAsync(Array.from(pathsToDelete), 100, async (pathToDelete) => {
+      try {
+        FileSystem.deleteFile(pathToDelete, { throwIfNotExists: true });
+        logger.terminal.writeVerboseLine(`Deleted "${pathToDelete}"`);
+        deletedFiles++;
+      } catch (error) {
+        if (FileSystem.exists(pathToDelete)) {
+          FileSystem.deleteFolder(pathToDelete);
+          logger.terminal.writeVerboseLine(`Deleted folder "${pathToDelete}"`);
+          deletedFolders++;
+        }
+      }
+    });
+
+    if (deletedFiles > 0 || deletedFolders > 0) {
+      logger.terminal.writeLine(`Deleted ${deletedFiles} files and ${deletedFolders} folders`);
+    }
+  }
+
+  private async _resolvePathAsync(globPattern: string, buildFolder: string): Promise<string[]> {
+    if (globEscape(globPattern) !== globPattern) {
+      const expandedGlob: string[] = await LegacyAdapters.convertCallbackToPromise(glob, globPattern, {
```

Comment:
I'd like to mention once again that globs are a performance killer. 

Every task that does this will potentially do its own brute-force enumeration of the entire directory tree, testing each path against its own particular glob expression.  This is going to add up over time.

Longer term we really should consider some other options:
- Due the crawl once and cache it in memory, then let the globs run against that in-memory tree
- Replace the [glob](https://www.npmjs.com/package/glob) package with some other package that is designed for performance
- Find a way to asynchronously test many globs simultaneously during a single crawl of the tree
- Avoid using globs; try to use specific folders instead

Reply:
Right now we're using this to clean up old generated files like the `src/**/*.scss.ts` files that `gulp-core-build-sass` created/creates. Perhaps that should be in its own bespoke plugin, and we only allow static paths here? That'd be a breaking change for our internal use of Heft, but it wouldn't be too bad to work around it.

Reply:
This is off topic from your PR obviously.

But maybe there are two initiatives we should think about pursuing:

1. We should replace the naive `glob` package with a heavily optimized glob API that runs against filesystem cache. It would be based around the assumption that the directory tree does not change except for changes made by Heft itself. I think this is similar to your cache optimization for the TypeScript compiler.  Maybe the idea can be rolled together into a family of APIs that all work against an in-memory cache of the directory tree.

2. We should design our config files and directory layouts to avoid the need for globs. Where we provide them, they are a last resort, and maybe a separate option.  For example, always offer two settings `pathsToDelete` and `globsToDelete`, with a comment warning people that the glob version will be slower.